### PR TITLE
mimxrt: Add the machine.spi() class to machine.

### PR DESCRIPTION
**Pull request by robert-hh**
This class supports both Controller (Master) an Peripheral (Slave) mode. In order to support Peripheral mode, two keyword arguments have been added:

mode=SPI.CONTROLLER or SPI.PERIPHERAL or SPI.MASTER or SPI.SLAVE
timeout=xxxx   Timeout to wait for a request. On timeout, an exception is raised

Further options:
drive=n  with n being between 1 and 6 or PIN.POWER_1 to PIN.POWER_6. Since the pins used
        by the SPI are fixed, no Pin settings can be made. Thus the drive option is added allowing
        to control ringing and crosstalk on the connection.
gap=nnnnn  The time between sent data items in a frame given in ns. Default is 2 clock cycles.

SPI device numbers start at 0, to comply with the Pin-Out of the Teensy boards.
With the configured clock frequency the fastest baud rate is 33MHz. For messages longer  16 bytes DMA is used. The class uses the existing frame work with extmod/machine_spi.c. I did not want to create a separate class for that.

There are still two dark spots on the implementation:
- Using DMA the data cache is disabled for the short transfer time. Otherwise data access was not reliable. Other less intrusive methods I tried did not work. Suggestions welcome.
- For short messages (<= 16 bytes) in slave mode of the MIMXRT1020_EVK board, write requests by a master are not detected and served in about 1 of 1000 cases, independent from baud rate and request frequency. DMA transfer is reliable. And no problem with the Teensy boards.

### CODE REVIEW

**Comment by robert-hh** [comment]
It seems that the cache is not the big problem. I made a test, how long the transfer set-up take with cache enabled and disabled. Note: The cache is disabled only for the last part of the set-up and the actual transfer. The test consists of writing a dynamically created message 10 time in a tight loop. The results.
```
Loop #   with cache   w/o cache
 1          80µs         80µs
2-10        30µs         70µs
```
The 70µs number has little variation, while the 30µs is an average with variation between 10 and 50 µs. So what you see essentially is the effect of cache. The cache is not big. If the code does not run in a tight loop, it is more likely that there is a cache miss in a call to spi.write(), and therefore there is no difference between the two options.

**Comment by dpgeorge** [comment]
I did not look at this in detail, but the first thing I'd say (as I mentioned elsewhere) is that slave/peripheral mode does not yet have an API defined, and eventually I think it should be something like `machine.SPIPeripheral` for peripheral mode.  So I would suggest to remove the peripheral support from this PR (of course don't discard the code, it will be useful later!).

**Comment by robert-hh** [comment]
I can do that, even if it was very helpful for testing, allowing a loop test. I do not favor the idea of having just another spi class. My impression is, that the only thing needed may be a return signal form the transfer call, like it exists for UART or I2C, where it is used for error reporting and telling  the number of bytes transferred.

**Comment by robert-hh** [comment]
The peripheral mode is removed, including the keyword parameters related to it.

File: extmod/machine_spi.h [hunk 1]
```diff
@@ -30,6 +30,12 @@
 #include "py/mphal.h"
 #include "drivers/bus/spi.h"
 
+// if a port didn't define MASTER/SLAVE constants then provide them
+#ifndef MICROPY_PY_MACHINE_SPI_MASTER
+#define MICROPY_PY_MACHINE_SPI_MASTER (1)
+#define MICROPY_PY_MACHINE_SPI_SLAVE  (0)
+#endif
```
Comment (dpgeorge) [inline extmod/machine_spi.h:37]: these can go

File: ports/mimxrt/board_init.c [hunk 2]
```diff
@@ -43,6 +43,7 @@
 volatile uint32_t systick_ms = 0;
 
 const uint8_t dcd_data[] = { 0x00 };
+void BOARD_ConfigMPU(void);
```
Comment (dpgeorge) [inline ports/mimxrt/board_init.c:46]: this is not needed

File: ports/mimxrt/dma_channel.c [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include "dma_channel.h"
+
+// List of channel flags: true: channel used, false: channel available
+static bool channel_list[32] = { true, true, true, true, false, false, false, false,
+                                 false, false, false, false, false, false, false, false,
+                                 false, false, false, false, false, false, false, false,
+                                 false, false, false, false, false, false, false, false };
+
+// allocate_channel(): retrieve an available channel. Return the number or -1
+int allocate_DMA_channel(void) {
```
Comment (dpgeorge) [inline ports/mimxrt/dma_channel.c:36]: the upper case "DMA" looks out of place, doesn't match other code styling; `allocate_dma_channel` would be better

File: ports/mimxrt/dma_channel.c [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        if (channel_list[i] == false) { // Channel available
+            channel_list[i] = true;
+            return i;
+        }
+    }
+    return -1;
+}
+
+// free_channel(n): Declare channel as free
+void free_DMA_channel(int n) {
```
Comment (dpgeorge) [inline ports/mimxrt/dma_channel.c:47]: similar to above, make lower case "DMA"

File: ports/mimxrt/machine_spi.c [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+#include "fsl_lpspi.h"
+#include "fsl_lpspi_edma.h"
+
+#define DEFAULT_SPI_BAUDRATE    (1000000)
+#define DEFAULT_SPI_POLARITY    (0)
+#define DEFAULT_SPI_PHASE       (0)
+#define DEFAULT_SPI_BITS        (8)
+#define DEFAULT_SPI_FIRSTBIT    (kLPSPI_MsbFirst)
+#define DEFAULT_SPI_MODE        (MICROPY_PY_MACHINE_SPI_MASTER)
```
Comment (dpgeorge) [inline ports/mimxrt/machine_spi.c:48]: this line isn't needed anymore

File: ports/mimxrt/machine_spi.c [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+STATIC const uint8_t spi_index_table[] = MICROPY_HW_SPI_INDEX;
+STATIC LPSPI_Type *spi_base_ptr_table[] = LPSPI_BASE_PTRS;
+static const iomux_table_t iomux_table[] = {
+    IOMUX_TABLE_SPI
+};
+
+static const char *firstbit_str[] = {"MSB", "LSB"};
+
+#define MICROPY_HW_SPI_NUM     (sizeof(spi_index_table) / sizeof(spi_index_table)[0])
```
Comment (dpgeorge) [inline ports/mimxrt/machine_spi.c:79]: you can use `MP_ARRAY_SIZE(spi_index_table)`

File: ports/mimxrt/machine_spi.c [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+mp_obj_t machine_spi_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_id, ARG_baudrate, ARG_polarity, ARG_phase, ARG_bits, ARG_firstbit, ARG_gap, ARG_drive };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id,       MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_baudrate, MP_ARG_INT, {.u_int = DEFAULT_SPI_BAUDRATE} },
+        { MP_QSTR_polarity, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_POLARITY} },
+        { MP_QSTR_phase,    MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_PHASE} },
+        { MP_QSTR_bits,     MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_BITS} },
+        { MP_QSTR_firstbit, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_FIRSTBIT} },
+        { MP_QSTR_gap,      MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
```
Comment (dpgeorge) [inline ports/mimxrt/machine_spi.c:130]: this is a new argument, maybe it can be called `gap_ns` to emphasise the units?

**Review by dpgeorge — CHANGES_REQUESTED** [review]

**Comment by robert-hh** [comment]
Thanks for the comments. All changes have been implemented.

**Review by dpgeorge — APPROVED** [review]

**Comment by robert-hh** [comment]
Thanks for approving. Should I rebase & squash the PR?

**Comment by dpgeorge** [comment]
> Thanks for approving. Should I rebase & squash the PR?

No need, done during merge.

Merged in f45412793e78cd5013b83b8de5219183eb53eca5

**Pull request closed without merging**